[core,lua] Action packet rework

sruon opened the pull request:
<!-- Remove space and place 'x' mark between square [] brackets or click the checkbox after saving to affirm the following points: -->
<!-- (it should look like this: - [x] I have ...) -->
**_I affirm:_**
- [x] I understand that if I do not agree to the following points by completing the checkboxes my PR will be ignored.
- [x] I understand I should leave resolving conversations to the LandSandBoat team so that reviewers won't miss what was said.
- [x] I have read and understood the [Contributing Guide](https://github.com/LandSandBoat/server/blob/base/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/LandSandBoat/server/blob/base/CODE_OF_CONDUCT.md).
- [x] I have _**tested my code and the things my code has changed**_ since the last commit in the PR and will test after any later commits.

## What does this pull request do?

**I have split the work into several commits for ease of review and history purposes but they do not compile on their own and some changes may be split across several commits.**

### Gameplay changes
- ~~Made some tweaks to RangedState to not cancel RAs if target is out of range~~
  - I've been told this was supposedly working before but the test was incredibly flaky without this so I'm not sure, I'll be retesting
- Self-centered AoEs no longer add the "primary target" by default
  - For damage/enfeeb moves: This was causing the primary target to always be hit even if they were not in range, but these skills CAN miss with a very specific error message "**No valid target in area of effect.**"
- Defenders should now more reliably visibly "recoil" when hit by a physical attack
- Konzen-Ittai should use the right animation... ~~maybe, I haven't tested it yet~~

### xi_test enhancements
- Adds a helper to execute a skillchain
  - Attempting to engage and then simulate running weaponskills proved just a tiny bit too complex for xi_test currently, specifically to respect the timings... so we're gonna cheat for now
- Adds a helper to retrieve test character received action packets

### Action packet GP_SERV_COMMAND_BATTLE2
- The packet constructor gets rid of all the magic values and is clearly documented with bit counts and fields matching atom0s various documentations.
  - pack() is used to bitpack an action to send to the client
  - unpack() is used to reverse for test purposes

### action_t
- Deprecated several existing enums in favor of new ones matching XiPackets docs (SPECEFFECT, REACTION, MODIFIER etc)
- Several semantic changes were made
  - The layering remains the same BUT the structs have been renamed to clearly indicate how they slot in the packet
    - Before: **action > list > target**
    - Now: **action > target > result** 
  - **reaction** is gone in favor of **resolution**: what the action resolved to (hit, miss, parry, block, guard)
  - **speceffect** is gone in favor of **info**, **hitDistortion** and **knockback** (the latter two are known as a single field **scale** in the packet but for all intents and purposes they can be treated separately)
- The lua bindings expose **info**, **hitDistortion** and **knockback** separately
- The lua bindings also expose **recordDamage**, taking a target, attack type, damage and crit
  - This is the preferred way to report damages as it does set the **hit distortion** and various bits in **info** transparently.
 
### Action interruptions
- A bunch of made up interruptions were removed and should absolutely not be reintroduced without a capture
- All interrupts were centralized in a dedicated `ActionInterrupts` namespace
  - These are all retail accurate down to the bit
  - ~~**Note: I still need to tweak a few code paths to not emit extra packets**~~

### AI
- action_t building has been reworked to make use of brace-init as much as possible
- Removed all magic values that did not make sense or were properly handled later (i.e. hit distortion/info)

### Tests
Introduces 60ish tests, all based on retail captures (a couple of TODOs here and there that are not overly critical)
 - Painfully aware some of the tests look pure garbage but xi_test needs a little bit of love first before I can refactor them

### Misc
- Removed ActionList dead code from BattleEntity

### Conclusion

**Very likely** to introduce **visual** bugs but the flow to resolve them is simple:
- Capture on retail
- Add the captured action packets as a new test
- Make changes until it passes!

I started working [on this web parser](https://sruon.github.io/parser/) that prints the format expected for testing - no where near done but it's more or less the format used by atom0s/actionparse

sruon committed:
Let RAs complete even if target is out of range

sruon committed:
Skillchain helper

sruon committed:
SPECEFFECT deprecated in favor of ActionInfo

sruon committed:
Deprecate various enums in favor of ActionXXX enums

sruon committed:
FourCC lua enum

sruon committed:
xi_test action packets helper

sruon committed:
GP_SERV_COMMAND_BATTLE2

sruon committed:
GP_SERV_COMMAND_BATTLE2 tests

sruon committed:
Reworked action structs

sruon committed:
Reworked Action Interrupts

sruon committed:
ActionList dead code

sruon committed:
Allow AoE moves centered on caster to miss if no target found

sruon committed:
Refactor skillchain/addEffects

Action skillchain helper

sruon committed:
rename action_t helpers

sruon committed:
Refactor how lua records action packet damage

sruon committed:
Refactor Zanshin check

sruon committed:
Refactor spikes

sruon commented on scripts/globals/ability.lua:
```diff
@@ -100,17 +100,12 @@ end
         end
 
-        action:param(defender:getID(), finaldmg)
+        action:recordDamage(defender, xi.attackType.PHYSICAL, finaldmg)
```
Need to check attackType here

sruon commented on scripts/globals/job_utils/dancer.lua:
```diff
@@ -257,7 +257,7 @@ end
 xi.job_utils.dancer.useStepAbility = function(player, target, ability, action, stepEffect)
     local stepInfoData     = actionInfo[ability:getID()]
-    local infoValue        = stepInfoData[0]
+    local infoValue        = stepInfoData[1]
```
This is a separate bugfix for DNC animations, I may have forgotten lua is 1-indexed in my previous PR.

sruon commented on scripts/globals/weaponskills.lua:
```diff
@@ -955,7 +954,7 @@ xi.weaponskills.takeWeaponskillDamage = function(defender, attacker, wsParams, p
         end
 
-        action:param(defender:getID(), math.abs(finaldmg))
+        action:recordDamage(defender, attack.type, math.abs(finaldmg)) -- TODO: Critical bit?
```
I swear I've seen WS set the critical hit bit but now it's all a blur, need to go retest.

sruon replied:
True Strike (forced crit)

<img width="414" height="146" alt="image" src="https://github.com/user-attachments/assets/3bd324ac-f0d2-4358-9078-481f3949dadf" />

Non-Crit Blade:Jin

<img width="510" height="161" alt="image" src="https://github.com/user-attachments/assets/9478e945-6d85-4feb-a1f6-de09d82a39df" />

Crit Blade: Jin
<img width="504" height="569" alt="image" src="https://github.com/user-attachments/assets/ca66a767-fd4b-49e2-a4b8-d8fc37adea44" />

sruon commented on scripts/globals/weaponskills.lua:
```diff
@@ -982,12 +981,6 @@ xi.weaponskills.takeWeaponskillDamage = function(defender, attacker, wsParams, p
 
     finaldmg = defender:takeWeaponskillDamage(attacker, finaldmg, attack.type, attack.damageType, attack.slot, primaryMsg, wsResults.tpHitsLanded * attackerTPMult, (wsResults.extraHitsLanded * 10) + wsResults.bonusTP, targetTPMult)
 
-    if wsResults.tpHitsLanded + wsResults.extraHitsLanded > 0 then
```
This block was a duplicate of a similar block above. But I'll re-check it specifically.

sruon commented on scripts/tests/packets/s2c/0x028_battle2/base.lua:
```diff
@@ -0,0 +1,255 @@
+---@type table<string, Battle2TestSuite>
+local testSuites =
+{
+    ['Abilities']        = require('scripts.tests.packets.s2c.0x028_battle2.abilities'),
```
Not overly fond of the fully qualified requires but I haven't found a clean way to do relative requires.

sruon commented on scripts/tests/packets/s2c/0x028_battle2/base.lua:
```diff
@@ -0,0 +1,255 @@
+}
+
+-- TODO: Replace with some higher level diff library
+local function diffPacket(actual, expected, prefix, phValues)
```
All of this will be replaced by a generic diff capability for xi_test, as soon as I figure out how to make that happen.

sruon commented on src/map/action/action.h:
```diff
@@ -0,0 +1,235 @@
+        }
+    }
+
+    void normalize()
```
Should probably move this to the .cpp

sruon commented on src/map/action/interrupts.h:
```diff
@@ -0,0 +1,59 @@
+class CBattleEntity;
+
+// Helpers to build specific action packet interruption flows
+// Baby chocobos will die if you make changes without a proper test with retail captures
```
Only half joking.

sruon commented on src/map/ai/helpers/targetfind.cpp:
```diff
@@ -464,7 +479,8 @@ bool CTargetFind::validEntity(CBattleEntity* PTarget)
     }
 
     // this is first target, always add him first
-    if (m_PTarget == nullptr)
```
This is what was adding the primary target regardless, even if they weren't in range.

sruon commented on src/map/ai/states/ability_state.cpp:
```diff
@@ -215,17 +207,6 @@ bool CAbilityState::CanUseAbility()
                 return false;
             }
 
-            // TODO: Remove this when all pet abilities are moved to PetSkill system.
```
Winter said this is unused, and I haven't seen anything going against that.

sruon commented on src/map/ai/states/ability_state.cpp:
```diff
@@ -261,35 +241,11 @@ bool CAbilityState::CanUseAbility()
 
         if (cancelAbility)
         {
-            // Only create a packet if the ability isn't instant
```
This is one of multiple former interruptions for abilities but there is not a single ability with a castTime > 0 now that Wyvern abilities are using the Pet skill code path.

sruon commented on src/map/ai/states/item_state.cpp:
```diff
@@ -255,12 +263,14 @@ void CItemState::TryInterrupt(CBattleEntity* PTarget)
     {
-        msg           = MSGBASIC_IS_PARALYZED;
+        ActionInterrupts::ItemParalyzed(m_PEntity, PTarget);
+        msg           = MSGBASIC_NONE; // The action packet already notifies.
```
Item state used to explicitly send message packets but the error messages are all part of the action packet!

sruon commented on src/map/ai/states/petskill_state.cpp:
```diff
@@ -64,31 +67,31 @@ CPetSkillState::CPetSkillState(CPetEntity* PEntity, uint16 targid, uint16 wsid)
+
+        m_PEntity->loc.zone->PushPacket(m_PEntity, CHAR_INRANGE, std::make_unique<GP_SERV_COMMAND_BATTLE2>(action));
+
+        // Wyverns immediately emit a skill interrupt packet.
```
Probably not the best location for this, but it will do for now.

sruon commented on src/map/ai/states/range_state.cpp:
```diff
@@ -134,47 +142,51 @@ bool CRangeState::Update(timer::time_point tick)
+            // Run CanUseRangedAttack but only for specific checks (ammo, animation state, etc)
+            // Don't let distance/LOS checks cause interrupts - those should produce misses
+            CanUseRangedAttack(static_cast<CBattleEntity*>(PTarget), true);
+        }
```
Now that I'm re-reading it, I'm not even sure it's required. I'll be retesting this block specifically.

sruon commented on src/map/entities/battleentity.cpp:
```diff
@@ -2032,33 +2029,41 @@ void CBattleEntity::OnCastFinished(CMagicState& state, action_t& action)
+            // Handle EFFECT_NONE - spell failed to apply
+            if (damage == EFFECT_NONE)
+            {
+                actionResult.resolution = ActionResolution::Miss;
```
This correctly sets the fields in the packet when a spell fails to take effect (i.e. Stoneskin on top of Stoneskin)

Super important to return EFFECT_NONE from the luas in such cases!

sruon commented on src/map/entities/battleentity.cpp:
```diff
@@ -2281,27 +2263,24 @@ void CBattleEntity::OnMobSkillFinished(CMobSkillState& state, action_t& action)
-            actionTarget.animation  = PSkill->getAnimationID();
-            actionTarget.reaction   = REACTION::MISS | REACTION::HIT | REACTION::GUARDED;
-            actionTarget.speceffect = SPECEFFECT::SELFAOE_MISS;
+            actionResult.animation  = PSkill->getAnimationID();
```
This block specifically handles Ruszor-style moves where the move does still go off even if no target is in range. The changes to Targetfind allows it to work correctly.

sruon commented on src/map/entities/battleentity.cpp:
```diff
@@ -2314,31 +2293,40 @@ void CBattleEntity::OnMobSkillFinished(CMobSkillState& state, action_t& action)
 
     bool first{ true };
+
+    // Process self first if present in target list (for self-buff abilities)
```
Removing the auto-assignment of the first target meant that the caster did not necessarily end up being the first target... so we're re-sorting it here. 

Targetfind should be rewritten to handle all of these cases correctly, it's only 500 lines long after all.

sruon commented on src/map/entities/charentity.cpp:
```diff
@@ -1696,7 +1637,7 @@ void CCharEntity::OnAbility(CAbilityState& state, action_t& action)
         if (this != PTarget && distance(this->loc.p, PTarget->loc.p) > PAbility->getRange())
         {
-            setActionInterrupted(action, PTarget, MSGBASIC_TOO_FAR_AWAY, 0);
+            // TODO: Is this relevant? Out of range is a BATTLE_MESSAGE, not an interrupt...
```
Shouldn't even get there since all abilities are instant cast and distance is checked well before this function is called

sruon commented on src/map/entities/charentity.cpp:
```diff
@@ -2023,39 +1942,20 @@ void CCharEntity::OnRangedAttack(CRangeState& state, action_t& action)
-        actionTarget.reaction        = REACTION::NONE;
-
+        ActionInterrupts::RangedParalyzed(this);
+        // TODO: The caller sends a useless empty packet
```
With the rework of interrupts, some of the code paths send an extra empty action packet... the client doesn't care but I need to fix it before we can merge this

sruon commented on src/map/entities/petentity.cpp:
```diff
@@ -562,53 +542,65 @@ void CPetEntity::OnPetSkillFinished(CPetSkillState& state, action_t& action)
+        }
+        else if (damage > 0 && PSkill->isDamageMsg())
+        {
+            // Check if Lua already set the damage via action:damage()
```
1000% garbage. Want to fix before merging.

Xaver-DaRed commented:
Godspeed

WinterSolstice8 commented on scripts/tests/packets/s2c/0x028_battle2/base.lua:
```diff
@@ -0,0 +1,255 @@
+                zone  = xi.zone.QUFIM_ISLAND,
+            })
+
+        mob = player.entities:moveTo(17293357)
```
probably should use name lookups for these in case of ID shifts

WinterSolstice8 commented on src/map/entities/charentity.cpp:
```diff
@@ -1474,7 +1460,8 @@ void CCharEntity::OnCastInterrupted(CMagicState& state, action_t& action, MSGBAS
         auto message = state.GetErrorMsg();
 
-        if (message && action.actiontype != ACTION_MAGIC_INTERRUPT) // Interrupt is handled elsewhere
+        // TODO: May need special handling if interrupt was handled elsewhere
```
"error" messages should probably be removed. States are weird and generate exceptions. Future problem though

sruon committed:
Konzen-Ittai named values

sruon committed:
Test messages, mob lookups

sruon committed:
Use ability attack type

sruon committed:
Proper action enums for RUN abilities

sruon committed:
Set critical hit bit on weaponskills

sruon committed:
action_t::normalize

sruon committed:
Set hit distortion on basic attacks

sruon committed:
Refine hit distortion ranges

sruon committed:
Set the crit flag and distortion on pet/mob skills

sruon committed:
Revert unnecessary fix for RA state

sruon committed:
Dont send empty action packets on interrupts

sruon committed:
Fix DNC/DRG scripts

sruon commented:
Apologies for the insane number of commits but I prefer to be very explicit in this massive PR. Feel free to squash.

Addressed all the issues I could think of:
- Konzen-Ittai uses values from a local table for the animations (+ retested)
- Used name lookups in tests instead of raw IDs
- xi.ability.takeDamage uses the passed attackType instead of defaulting to Physical
- Bugfix for RUN resolution / DRG jump info
- Confirmed retail sets critical hit bit on weaponskills and added support for it
- Made basic attacks use the recordDamage path so they properly set hit distortion
- Lowered thresholds for hit distortion to 0%/10%/20% after testing it some more
- Confirmed retail sets critical hit bit and distortion on petskills and mobskills and added support for it
  - Note: This introduces new bindings on CMobAbility/CPetAbility to support storing the values... this is absolutely not how it should be long term but it is a pattern similar to what's done for messages etc. 
  - The alternative was modifying 880 files and the whole thing is in the middle of being changed. I'll revisit at some point.
- Reverted the fix to ranged state, it turns out LSB was already correctly forcing misses on RAs and my tests were flaky for unrelated reasons
- Protected several code paths to avoid sending empty action packets when we back out early out of the states due to interrupts
-  Unrelated fix for DNC Desperate Flourish

I spent some time testing SMN/DNC/RUN/DRG/BST but I can't promise I have tested every single case. If issues come up we should be able to resolve them quickly with confidence thanks to the tests.

This is as good as it's gonna get for the time being.

sruon commented:
> 
> I noticed when I was working on mobskills that we don't expose action lua side for mobskills like we do for say, weaponskills.
> 
> In future, after we get mobskills sorted out, would the way to go be to add "action" to onMobWeaponSkill so we could set crits, blocks, etc when applicable in mobskills.lua?

Passing action is a very simple addition and I'm not sure why it wasn't done to begin with, but I'd go one step further and say these lua functions should absolutely not be concerned with what gets presented to the client and rather be strictly limited to calculating the outcome of the mobskill/ws/petskill.

What I mean specifically:
- Setting messages shouldn't be part of a skill
- Handling the damage to set absorb message if damage < 0 shouldn't be part of a skill
- Setting critical hit bit/defeated bit shouldn't be part of a skill
- Setting guarded/block/parry bit shouldn't be part of a skill
etc.

I think the way forward should be to have these functions return a table that represents the outcome of an ability such that

```lua
return 
{
  criticalHits    = 1,
  guardedHits = 1,
  missedHits  = 1,
  damage       = 1000,
  -- Anything else that makes sense
}
```

This will let us flow important information to core for other purposes and keep a clear separation of concerns between the two.
